Gems/Atom/RPI/Code/Source/RPI.Public/Pass/PassContainer.cpp

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include <Atom/RPI.Public/Pass/PassContainer.h>
+#include <Atom/RPI.Public/Pass/PassUtils.h>
+
+namespace AZ::RPI
+{
+    void PassContainer::EraseFromLists(AZStd::function<bool(const RHI::Ptr<Pass>&)> predicate)
+    {
+        erase_if(m_removePassList, predicate);
+        erase_if(m_buildPassList, predicate);
+        erase_if(m_initializePassList, predicate);
+    }
+
+    void PassContainer::ClearQueues()
+    {
+        m_removePassList.clear();
+        m_buildPassList.clear();
+        m_initializePassList.clear();
+    }
+
+    void PassContainer::RemovePasses()
+    {
+        AZ_PROFILE_SCOPE(RPI, "PassContainer::RemovePasses");
+
+        if (!m_removePassList.empty())
+        {
+            PassUtils::SortPassListDescending(m_removePassList);
+
+            for (const Ptr<Pass>& pass : m_removePassList)
+            {
+                pass->RemoveFromParent();
+            }
+
+            m_removePassList.clear();
+        }
+    }
+
+    void PassContainer::BuildPasses()
+    {
+        AZ_PROFILE_SCOPE(RPI, "PassContainer::BuildPasses");
+
+        m_passesChangedThisFrame = m_passesChangedThisFrame || !m_buildPassList.empty();
+
+        // While loop is for the event in which passes being built add more pass to m_buildPassList
+        while (!m_buildPassList.empty())
+        {
+            AZ_Assert(m_removePassList.empty(), "Passes shouldn't be queued removal during build attachment process");
+
+            // Copy the list and clear it, so that further passes can be queue during this phase
+            AZStd::vector< Ptr<Pass> > buildListCopy = m_buildPassList;
```

Comment:
This is the same logic as was already in the pass system. Not expecting different threads to add to the list now that the lists are per-pipeline, but within the same thread deeper in the callstack a ParentPass that's Building might create and thereby queue it's child pass for build. The passes themselves have a state enum to avoid duplicate calls to build/initialization.